### ISSUE TITLE
Make `GetOptionsExt` publicly usable

### ISSUE THREAD

**Issue by gruuya** [apache/arrow-rs#4525]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Ideally, to implement the new `ObjectStore::get_opts` for other object stores based on `reqwest` clients, we could simply call  `GetOptionsExt::with_get_options` on the `RequestBuilder`, as is done in the `object_store` crate itself. 

This is not currently possible (I assume inadvertently as the trait itself is public?), since the containing `client` mod is currently [private](https://github.com/apache/arrow-rs/blob/master/object_store/src/lib.rs#L245); 

**Describe the solution you'd like**
Make the trait public/exportable beyond the `object_store` crate.

**Describe alternatives you've considered**
Repeat the logic from `GetOptionsExt::with_get_options`.

**Additional context**
Over at Seafowl we have a [custom HTTP object store](https://github.com/splitgraph/seafowl/blob/main/src/object_store/http.rs) implementation for legacy reasons.

**Comment by tustvold** [comment]
These extension traits are intentionally private, as they're really just implementation details, that being said `GetOptionsExt` seems fairly harmless to make public

**Comment by Blizzara** [comment]
Hey @tustvold, somewhat related question - would you consider making the whole `client` mod public? We have our own custom ObjectStore implementation that's backed by an S3 api with some extras, but I think we could reuse most/all of the `client`. Alternatively I can also just make a copy of it for us, but wanted to check here first :)

**Comment by tustvold** [comment]
I don't think we could make it public in its current form, the API surface is simply too large and not really designed around API stability. 

I think probably we would need something like #6056 to provide a coherent contract, which we could then provide a default implementation of.
